sync_labels.py

```diff
@@ -0,0 +1,148 @@
+import sys
+import os
+import argparse
+from github import Github, GithubException
+
+"""
+Sync Github Labels
+
+Synchronize labels idempotently between two Github repositories.
+"""
+
+
+def sync_labels(src_repo, dest_repo, dry_run, delete):
+    """
+    Main subroutine: sync labels between source/destination Github repos
+
+    :param src_repo: full repo name for source of labels
+    :param dest_repo: full repo name for destination of label actions
+    :param dry_run: do a dry run, do not perform the actions
+    :param delete: delete labels that are in destination but not in source
+    """
+
+    # Get PyGithub objects representing repos
+    g = Github(get_access_token())
+    try:
+        s = g.get_repo(src_repo)
+    except GithubException:
+        raise Exception(f"Source repo is not a repo: {src_repo}")
+    try:
+        d = g.get_repo(dest_repo)
+    except GithubException:
+        raise Exception(f"Destination repo is not a repo: {dest_repo}")
+
+    # Create list of labels to add/delete
+    src_labels = list(s.get_labels())
+    src_label_names = set([j.name for j in src_labels])
+    dest_labels = list(d.get_labels())
+    dest_label_names = set([j.name for j in dest_labels])
+    dest_add = src_label_names - dest_label_names
+    dest_del = dest_label_names - src_label_names
+
+    # Show summary of what actions will be taken, get confirmation
+    print(f"The following labels will be created in {dest_repo}:")
+    for lab in sorted(list(dest_add)):
+        print(f" [+] {lab}")
+
+    if delete:
+        print(f"The following labels will be deleted in {dest_repo}:")
+        for lab in sorted(list(dest_del)):
+            print(f" [-] {lab}")
+
+    # Add new labels
+    for name in dest_add:
+        print(f"Creating label {name} in destination repo {dest_repo}...")
+        if dry_run:
+            print(f"Dry run successful!")
+            continue
+        label = s.get_label(name)
+        color = label.color
+        description = label.description
+        try:
+            d.create_label(name, color, description)
+            print(f"Success!")
+        except GithubException:
+            raise Exception(f"Error creating label {name} in destination repo {dest_repo}")
+
+    # Delete labels
+    if delete:
+        for name in dest_del:
+            print(f"Removing label {name} from destination repo {dest_repo}...")
+            if dry_run:
+                print(f"Dry run successful!")
+                continue
+            label = d.get_label(name)
+            try:
+                label.delete()
+                print(f"Success!")
+            except GithubException:
+                raise Exception(f"Error deleting label {name} from destination repo {dest_repo}")
+
+    print("Done.")
+
+
+def get_access_token():
+    """Grab github api token from env var"""
+    api_key = os.environ.get("GITHUB_API_KEY", None)
+    if api_key is None:
+        raise Exception("No env var GITHUB_API_KEY defined")
```

Comment:
Usually it's preferred to raise concrete exceptions instead of `Exception` directly. `ValueError` or `RuntimeError` are ones I use frequently in instances like this, but nothing beats spinning your own.

This comment applies to all uses of `Exception` in the script.

Reply:
addressed by 2927311